src/Compilers/Core/Portable/Compilation/IExpression.cs

```diff
@@ -0,0 +1,778 @@
+        /// </summary>
+        /// <param name="parameter">Parameter of the target method.</param>
+        /// <returns>Argument corresponding to the parameter.</returns>
+        IArgument ArgumentMatchingParameter(IParameterSymbol parameter);
+    }
+
+    /// <summary>
+    /// Kinds of invocations.
+    /// </summary>
+    public enum InvocationKind
+    {
+        /// <summary>
+        /// Virtual invocation.
+        /// </summary>
+        Virtual,
+        /// <summary>
+        /// Static or shared invocation with no instance value.
+        /// </summary>
+        Static,
+        /// <summary>
+        /// Non-virtual invocation with an instance value.
+        /// </summary>
+        NonVirtualInstance
+    }
+
+    /// <summary>
+    /// Represents an argument in a method invocation.
+    /// </summary>
+    public interface IArgument
+    {
+        /// <summary>
+        /// Kind of argument.
+        /// </summary>
+        ArgumentKind Kind { get; }
+        /// <summary>
+        /// Mode of argument.
+        /// </summary>
+        ArgumentMode Mode { get; }
+        /// <summary>
+        /// Value supplied for the argument.
+        /// </summary>
+        IExpression Value { get; }
+        /// <summary>
+        /// Conversion applied to the argument value passing it into the target method. Applicable only to In or Reference arguments.
+        /// </summary>
+        IExpression InConversion { get; }
+        /// <summary>
+        /// Conversion applied to the argument value after the invocation. Applicable only to Out or Reference arguments.
+        /// </summary>
+        IExpression OutConversion { get; }
+    }
+
+    /// <summary>
+    /// Kinds of arguments.
+    /// </summary>
+    public enum ArgumentKind
+    {
+        /// <summary>
+        /// Argument is specified positionally and matches the parameter of the same ordinality.
+        /// </summary>
+        Positional,
+        /// <summary>
+        /// Argument is specified by name and matches the parameter of the same name.
+        /// </summary>
+        Named,
+        /// <summary>
+        /// Argument becomes an element of an array that matches a trailing C# params or VB ParamArray parameter.
+        /// </summary>
+        ParamArray
+    }
+
+    /// <summary>
+    /// Modes of arguments.
+    /// </summary>
+    public enum ArgumentMode
+    {
+        /// <summary>
+        /// Argument is passed as an input value only.
+        /// </summary>
+        In,
+        /// <summary>
+        /// Argument is passed without an input value and is assigned a value by the time the invoked method returns.
+        /// </summary>
+        Out,
+        /// <summary>
+        /// Argument is passed by reference.
+        /// </summary>
+        Reference
+    }
+
+    /// <summary>
+    /// Represents a named argument.
+    /// </summary>
+    public interface INamedArgument : IArgument
```

Comment:
Do we need a separate interface for this?  I feel like 'Name' could simply be optional on IArgument.